chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,66 +39,179 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
     CHAINERX_NEVER_REACH();
 }
 
+namespace cuda_internal {
+
+std::unique_ptr<Chunk> Chunk::Split(size_t bytesize) {
+    CHAINERX_ASSERT(bytesize_ >= bytesize);
+    if (bytesize_ == bytesize) {
+        return nullptr;
+    }
+
+    std::unique_ptr<Chunk> remaining = std::make_unique<Chunk>(ptr(), bytesize, bytesize_ - bytesize);
+    bytesize_ = bytesize;
+
+    if (next_ != nullptr) {
+        remaining->SetNext(next_);
+        remaining->next()->SetPrev(remaining.get());
+    }
+    next_ = remaining.get();
+    remaining->SetPrev(this);
+
+    return remaining;
+}
+
+void Chunk::MergeWithNext() {
+    CHAINERX_ASSERT(next_ != nullptr);
+    bytesize_ += next_->bytesize();
+    if (next_->next() != nullptr) {
+        next_->next()->SetPrev(this);
+    }
+    next_ = next_->next();
+}
+
+}  // namespace cuda_internal
+
+namespace {
+using FreeBinsMap = std::map<size_t, cuda_internal::FreeList>;
+}
+
+// Pushes a chunk into an appropriate free list
+//
+// Not thread-safe
+void MemoryPool::PushIntoFreeList(std::unique_ptr<cuda_internal::Chunk> chunk) {
+    cuda_internal::FreeList& free_list = free_bins_[chunk->bytesize()];
+    free_list.emplace_back(std::move(chunk));
+}
+
+void MemoryPool::CompactFreeBins(FreeBinsMap::iterator it_start, FreeBinsMap::iterator it_end) {
+    auto it_start_rev = std::make_reverse_iterator(it_start);
+    it_start = std::find_if(it_start_rev, free_bins_.rend(), [](const FreeBinsMap::value_type& p) { return !p.second.empty(); }).base();
```

Comment:
`p.second` is not obvious on its type. How about aliasing to FreeList as a local variable as we do in ranged-for loop.

Reply:
Also, I feel `pair` improves readability than `p`.